volatility3/framework/plugins/windows/mftscan.py

```diff
@@ -189,3 +190,158 @@ def run(self):
+        ):
+            with contextlib.suppress(exceptions.PagedInvalidAddressException):
+                mft_record = self.context.object(
+                    mft_object, offset=offset, layer_name=layer.name
+                )
+                # We will update this on each pass in the next loop and use it as the new offset.
+                attr_base_offset = mft_record.FirstAttrOffset
+
+                attr_header = self.context.object(
+                    header_object,
+                    offset=offset + attr_base_offset,
+                    layer_name=layer.name,
+                )
+
+                # There is no field that has a count of Attributes
+                # Keep Attempting to read attributes until we get an invalid attr_header.AttrType
+                file_name = "N/A"
+                is_ads = False
+                    # The First $DATA Attr is the 'principal' file itself not the ADS
+                while attr_header.AttrType.is_valid_choice:
+                    # Offset past the headers to the attribute data
+                    attr_data_offset = (
+                        offset
+                        + attr_base_offset
+                        + self.context.symbol_space.get_type(
+                            attribute_object
+                        ).relative_child_offset("Attr_Data")
+                    )
+
+                    if attr_header.AttrType.lookup() == "FILE_NAME":
+                        attr_data = self.context.object(
+                            fn_object, offset=attr_data_offset, layer_name=layer.name
+                        )
+                        file_name = attr_data.get_full_name()
+                    
+
+                    # DATA Attribute (can be ADS or not)
+                    if attr_header.AttrType.lookup() == "DATA":
+                        if is_ads:
+                            if not attr_header.NonResidentFlag:
+                                # Resident files are the most interesting.
+                                if attr_header.NameLength > 0:
+                                    attr_name_offset = (
+                                        offset
+                                        + attr_base_offset
+                                        + attr_header.NameOffset
+                                    )
+                                    ads_name = self._context.layers[layer.name].read(
+                                        attr_name_offset, attr_header.NameLength*2 , pad=True
+                                    ).decode('utf-16')
+                                    attr_content_offset = (
+                                        offset
+                                        + attr_base_offset
+                                        + attr_header.ContentOffset
+                                    )                                    
+                                    content = self._context.layers[layer.name].read(
```

Comment:
This also feels a little nasty (I'm not keen on plugins doing direct layer reads), but I'm not sure how to do it more cleanly and this does seem to be unstructured binary data that just needs reading.  I think an array of bytes would just make this uglier, so probably best to keep it as it is.

Reply:
I agree, I was also looking for the best way. Looked at other plugins, seems like netstat for example is doing something similar, I figured it might be accepted like that.

```python3
        part_table = context.object(
            net_symbol_table + constants.BANG + "_PARTITION_TABLE",
            layer_name=layer_name,
            offset=part_table_addr,
        )
        part_count = int.from_bytes(
            context.layers[layer_name].read(tcpip_module_offset + part_count_symbol, 1),
            "little",
        )
```

Reply:
Yeah, I'd leave it as it is.  It probably is the right way of doing it, it just gave me the creeps so I thought I'd mention it...  5;P